src/com/sony/simdetect/SimDetectService.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019-2020 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Rewritten in Kotlin by Pavel Dubrova <pashadubrova@gmail.com>
+ */
+
+package com.sony.simdetect
+
+import android.app.Service
+import android.content.Context
+import android.content.Intent
+import android.os.Handler
+import android.os.IBinder
+import android.os.Looper
+import android.os.Message
+import android.os.UEventObserver
+
+import com.android.internal.R
+import com.android.internal.telephony.uicc.UiccSlot
+
+class SimDetectService : Service() {
+    private val Lock = Any()
+
+    private val SimDetectEventObserver = object : UEventObserver() {
+        override fun onUEvent(event: UEvent) {
+            synchronized(Lock) {
+                val switchState = event.get("SWITCH_STATE")
+                if (SIM_REMOVED.equals(switchState)) {
+                    promptForRestart(false)
+                } else if (SIM_INSERTED.equals(switchState)) {
+                    promptForRestart(true)
+                }
+            }
+        }
+    }
+
+    override fun onCreate() {
+        val isHotSwapSupported = getResources().getBoolean(R.bool.config_hotswapCapable)
+        if (!isHotSwapSupported) {
+            SimDetectEventObserver.startObserving("SWITCH_NAME=sim_detect")
+        }
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        return START_STICKY
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+    }
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return null
+    }
+
+    private fun promptForRestart(isAdded: Boolean) {
+        Handler(Looper.getMainLooper()).post({
+            val uiccSlot = UiccSlot(this@SimDetectService, false)
+            uiccSlot.sendMessage(uiccSlot.obtainMessage(
+                    if (isAdded) EVENT_CARD_ADDED else EVENT_CARD_REMOVED, null))
+        })
+    }
+
+    companion object {
```

Comment:
The `companion object` is not necessary here (neither in `BootCompletedReceiver`).

https://kotlinlang.org/docs/reference/properties.html
> Properties the value of which is known at compile time can be marked as compile time constants using the const modifier.

Where "compile time constants" implies that the value is static.

https://blog.egorand.me/where-do-i-put-my-constants-in-kotlin/
That's a nice read on const handling, and shows a `const` marked value will indeed get compiled to the java equivalent of `public/private static final String foo = "bar"`.